ouroboros-consensus/test/consensus-test/Test/Consensus/MiniProtocol/ObjectDiffusion/Smoke.hs

```diff
@@ -0,0 +1,315 @@
+
+newtype SmokeObjectId = SmokeObjectId Int
+  deriving (Eq, Ord, Show, NoThunks, Arbitrary)
+
+newtype SmokeObject = SmokeObject {smokeObjectId :: SmokeObjectId}
+  deriving (Eq, Ord, Show, NoThunks, Arbitrary)
+
+newtype SmokeObjectPool m = SmokeObjectPool (StrictTVar m [SmokeObject])
+
+newObjectPool :: MonadSTM m => [SmokeObject] -> m (SmokeObjectPool m)
+newObjectPool initialPoolContent = SmokeObjectPool <$> uncheckedNewTVarM initialPoolContent
+
+makeObjectPoolReader ::
+  MonadSTM m => SmokeObjectPool m -> ObjectPoolReader SmokeObjectId SmokeObject Int m
+makeObjectPoolReader (SmokeObjectPool poolContentTvar) =
+  ObjectPoolReader
+    { rdrGetObjectId = smokeObjectId
+    , objectPoolGetSnapshot = poolContentToSnapshot <$> readTVar poolContentTvar
+    , objectPoolZeroTicketNo = -1 -- objectPoolObjectIdsAfter uses strict comparison, and first index is 0.
+    }
+ where
+  poolContentToSnapshot :: [SmokeObject] -> ObjectPoolSnapshot SmokeObjectId SmokeObject Int
+  poolContentToSnapshot poolContent =
+    ObjectPoolSnapshot
+      { objectPoolObjectsAfter = \minIndex -> do
+          (index, smokeObject) <- zip [0 ..] poolContent
+          guard (index > minIndex)
+          return (smokeObject, index, SizeInBytes 0) -- REVIEW: 0?
+      , objectPoolHasObject = \objectId -> any ((== objectId) . smokeObjectId) poolContent
+      }
+
+makeObjectPoolWriter ::
+  MonadSTM m => SmokeObjectPool m -> ObjectPoolWriter SmokeObjectId SmokeObject m
+makeObjectPoolWriter (SmokeObjectPool poolContent) =
+  ObjectPoolWriter
+    { wrGetObjectId = smokeObjectId
+    , objectPoolAddObjects = \objects -> do
+        atomically $ modifyTVar poolContent (++ objects)
+        return ()
+    }
+
+mkMockPoolInterfaces ::
+  MonadSTM m =>
+  [SmokeObject] ->
+  m
+    ( ObjectPoolReader SmokeObjectId SmokeObject Int m
+    , ObjectPoolReader SmokeObjectId SmokeObject Int m
+    , ObjectPoolWriter SmokeObjectId SmokeObject m
+    )
+mkMockPoolInterfaces objects = do
+  outboundPool <- newObjectPool objects
+  inboundPool <- newObjectPool []
+
+  let outboundPoolReader = makeObjectPoolReader outboundPool
+      inboundPoolReader = makeObjectPoolReader inboundPool
+      inboundPoolWriter = makeObjectPoolWriter inboundPool
+
+  return (outboundPoolReader, inboundPoolReader, inboundPoolWriter)
+
+{-------------------------------------------------------------------------------
+  Main properties
+-------------------------------------------------------------------------------}
+
+-- Protocol constants
+
+numObjectIdsToAck :: NumObjectIdsToAck
+numObjectIdsToAck = NumObjectIdsToAck 10
+
+nodeToNodeVersion :: NodeToNodeVersion
+nodeToNodeVersion = NodeToNodeV_14
+
+prop_smoke_init_inbound :: ListWithUniqueIds SmokeObject idTy -> Property
+prop_smoke_init_inbound (ListWithUniqueIds objects) =
+  prop_smoke_object_diffusion objects mkOutboundAsync mkInboundAsync (mkMockPoolInterfaces objects)
+ where
+  mkOutboundAsync outbound outboundChannel tracer =
+    async $
```

Comment:
FYI: For simple tests like this one, using `async` is fine; but in general, it suffers from resource leakage (as nothing makes sure that this thread is killed). We usually use [`forkLinkedThread`](https://hackage-content.haskell.org/package/resource-registry-0.1.1.0/docs/Control-ResourceRegistry.html#v:forkLinkedThread) as a safe alternative in the actual code as well as in more complicated tests.